### PR TITLE
[compiler] Add !cudnn.tensor type with a support for dynamic shapes

### PR DESCRIPTION
**Pull request by ezhulenev**
Add `!cudnn.tensor` type with a goal to replace `!cudnn.tensor_descriptor`

1. Add common tensor layouts as an enum attribute (NHWC, NCHW, etc...). See https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#tensors-layouts
2. Add support for dynamic shapes by defining strides as an affine map for non-standard layouts
3. Remove memory alignment from the type, as it is a run time property of allocated buffers

@chsigg @frgossen

### CODE REVIEW

File: compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/IR/CUDNNTypes.cpp [hunk 1]
```diff
@@ -8,34 +8,42 @@
...
 }
 
 static void printDimensionList(AsmPrinter &printer, ArrayRef<int64_t> dims,
                                Type type) {
-  llvm::interleave(dims, printer.getStream(), "x");
-  printer << 'x' << type;
+  auto print = [&](int64_t dim) {
+    ShapedType::isDynamic(dim) ? printer << "?" : printer << dim;
+  };
+  llvm::interleave(dims, printer.getStream(), print, "x");
```
Comment (chsigg) [inline compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/IR/CUDNNTypes.cpp:40]: Total nit, but wouldn't

`for (auto dim : dims) printer << dim << "x";`

be easier to read?

File: compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/IR/CUDNNTypes.cpp [hunk 2]
```diff
@@ -45,3 +53,96 @@ void CUDNNDialect::registerTypes() {
...
+//===----------------------------------------------------------------------===//
+
+namespace openxla::compiler::nvgpu::cudnn {
+
+bool TensorType::isOpaque() {
+  return getShape().empty() && !getElementType() && !getLayout() &&
+         !getStrides();
+}
+
+LogicalResult TensorType::verify(function_ref<InFlightDiagnostic()> emitError,
```
Comment (chsigg) [inline compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/IR/CUDNNTypes.cpp:68]: Would it make sense to cover these errors by tests?

File: tools/openxla-runner.c
Comment (chsigg) [inline tools/openxla-runner.c:1]: What are these changes for?

**Review by chsigg — APPROVED** [review]

File: tools/openxla-runner.c
Comment (ezhulenev) [reply tools/openxla-runner.c:1]: not related, just to fix a compilation issue with latest iree core

File: compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/IR/CUDNNTypes.cpp (hunk 2 above)
Comment (ezhulenev) [reply compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/IR/CUDNNTypes.cpp:68]: It triggers assertion error and crashes. I'll try to add them later once I'll find how to do it without crashing the tool.

**Pull request merged**